movement/watch_faces/complication/totp_face_lfs.c

```diff
@@ -0,0 +1,231 @@
+            num_totp_records += 1;
+        } else {
+            printf("TOTP missing secret: %s\n", line);
+        }
+    }
+}
+
+void totp_face_lfs_setup(movement_settings_t *settings, uint8_t watch_face_index, void ** context_ptr) {
+    (void) settings;
+    (void) watch_face_index;
+    if (*context_ptr == NULL) {
+        *context_ptr = malloc(sizeof(totp_lfs_state_t));
+        totp_face_lfs_read_file(TOTP_FILE);
+    }
+}
+
+void totp_face_lfs_activate(movement_settings_t *settings, void *context) {
+    (void) settings;
+    memset(context, 0, sizeof(totp_lfs_state_t));
+    totp_lfs_state_t *totp_state = (totp_lfs_state_t *)context;
+
+    if (num_totp_records > 0) {
+        totp_state->current_index = 0;
+        totp_state->steps = 0;  // Force getting a new code in the loop.
+        totp_state->timestamp = watch_utility_date_time_to_unix_time(watch_rtc_get_date_time(), movement_timezone_offsets[settings->bit.time_zone] * 60);
+        TOTP(totp_records[0].secret, totp_records[0].secret_size, totp_records[0].period);
+    } else {
+        watch_display_string("NA00000000", 0);
+    }
+}
+
+bool totp_face_lfs_loop(movement_event_t event, movement_settings_t *settings, void *context) {
+    (void) settings;
+
+    totp_lfs_state_t *totp_state = (totp_lfs_state_t *)context;
+    char buf[14];
+    uint8_t valid_for;
+    div_t result;
+    uint8_t index = totp_state->current_index;
+
+    switch (event.event_type) {
+        case EVENT_TICK:
+            if (num_totp_records == 0) {
+                break;
+            }
+
+            totp_state->timestamp++;
+            // fall through
+        case EVENT_ACTIVATE:
+            if (num_totp_records == 0) {
+                break;
+            }
+
+            result = div(totp_state->timestamp, totp_records[index].period);
+            if (result.quot != totp_state->steps) {
+                totp_state->current_code = getCodeFromTimestamp(totp_state->timestamp);
+                totp_state->steps = result.quot;
+            }
+            valid_for = totp_records[index].period - result.rem;
+            sprintf(buf, "%c%c%2d%06lu", totp_records[index].label[0], totp_records[index].label[1], valid_for, totp_state->current_code);
+
+            watch_display_string(buf, 0);
+            break;
+        case EVENT_MODE_BUTTON_UP:
+            movement_move_to_next_face();
+            break;
+        case EVENT_LIGHT_BUTTON_DOWN:
+            movement_illuminate_led();
+            break;
+        case EVENT_TIMEOUT:
+            movement_move_to_face(0);
+            break;
+        case EVENT_ALARM_BUTTON_UP:
+            if (num_totp_records == 0) {
+                break;
+            }
+            totp_state->current_index = (index + 1) % num_totp_records;
+            TOTP(totp_records[totp_state->current_index].secret, totp_records[totp_state->current_index].secret_size, totp_records[totp_state->current_index].period);
```

Comment:
Having used the TOTP face and understood more about what's going on, it would be good to do the display here. i.e. probably pull the ACTIVATE code into a display function, and call is from TICK/ACTIVATE/ALARM_BUTTON_UP.

Reply:
Def agree; this works, but lags by up to one second while it waits for the next tick.